test/extensions/transport_sockets/tls/tls_throughput_test.cc

```diff
@@ -0,0 +1,173 @@
+
+  bool handshake_success = false;
+  for (int i = 0; i < 50; i++) {
+    int client_err = SSL_do_handshake(client_ssl);
+    int server_err = SSL_do_handshake(server_ssl);
+    if (client_err == 1 && server_err == 1) {
+      handshake_success = true;
+      break;
+    }
+    int err = SSL_get_error(server_ssl, server_err);
+    switch (err) {
+    case SSL_ERROR_WANT_READ:
+    case SSL_ERROR_WANT_WRITE:
+      continue;
+    default:
+      drainErrorQueue();
+      PANIC("Unexpected error during handshake");
+    }
+
+    ENVOY_LOG_MISC(error, "client_err {} server_err {}", client_err, server_err);
+  }
+
+  RELEASE_ASSERT(handshake_success, "handshake completed successfully");
+
+  static uint8_t read_buf[1024 * 1024];
+
+  auto full_linearize = state.range(0);
+  unsigned short_slice_size = state.range(1);
+  unsigned num_short_slices = state.range(2);
+
+  uint64_t bytes_written = 0;
+  for (auto _ : state) {
+    state.PauseTiming();
+
+    // Empty out the read side to make space for the writes.
+    while (SSL_read(server_ssl, read_buf, sizeof(read_buf)) > 0)
+      ;
+
+    std::string send(short_slice_size, 'a');
+    Buffer::OwnedImpl write_buf;
+    for (unsigned i = 0; i < num_short_slices; i++) {
+      if (short_slice_size > 0) {
+        appendSlice(write_buf, short_slice_size);
+      }
+    }
+    appendSlice(write_buf, 16384 - (num_short_slices * short_slice_size));
+    RELEASE_ASSERT(write_buf.length() == 16384,
+                   fmt::format("expected length 16384, got {}", write_buf.length()));
+    RELEASE_ASSERT(write_buf.getRawSlices().size() == (num_short_slices + 1),
+                   fmt::format("buffer number of slices expected {}, got {}", num_short_slices + 1,
+                               write_buf.getRawSlices().size()));
+
+    // Append many full-sized slices, the same manner that an envoy socket read would.
+    for (unsigned i = 0; i < 10; i++) {
+      auto start_size = write_buf.length();
+      Buffer::RawSlice slices[2];
+      auto num_slices = write_buf.reserve(16384, slices, 2);
+      for (unsigned i = 0; i < num_slices; i++) {
+        memset(slices[i].mem_, 'a', slices[i].len_);
+      }
+      write_buf.commit(slices, 2);
+      RELEASE_ASSERT(write_buf.length() - start_size == 16384, "correct reserve/commit");
+    }
+
+    bytes_written += write_buf.length();
+
+    state.ResumeTiming();
+    uint32_t num_writes = 0;
+    while (write_buf.length() > 0) {
+      void* mem;
+      size_t len = std::min<uint64_t>(write_buf.length(), 16384);
+      if (full_linearize) {
+        mem = write_buf.linearize(len);
+      } else {
+        auto slice = write_buf.maybeLinearize(len, 4096);
+        mem = slice.mem_;
+        len = slice.len_;
+      }
+
+      err = SSL_write(client_ssl, mem, len);
+      RELEASE_ASSERT(err == static_cast<int>(len), "SSL_write");
+      write_buf.drain(len);
+      num_writes++;
+    }
+
+    state.counters["writes_per_iteration"] = num_writes;
```

Comment:
could we capture a count on the number of times linearize did something?

Reply:
Good idea!

```
----------------------------------------------------------------------------------
Benchmark                        Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------
testThroughput/0/0/0           330 us          328 us            1 num_times_linearize_did_something=10 throughput=548.771M/s writes_per_iteration=11
testThroughput/1/0/0           276 us          275 us            1 num_times_linearize_did_something=10 throughput=655.961M/s writes_per_iteration=11
testThroughput/0/1/1           285 us          285 us            1 num_times_linearize_did_something=10 throughput=633.111M/s writes_per_iteration=12
testThroughput/1/1/1           342 us          341 us            1 num_times_linearize_did_something=11 throughput=528.143M/s writes_per_iteration=11
testThroughput/0/128/1         290 us          289 us            1 num_times_linearize_did_something=9 throughput=623.632M/s writes_per_iteration=12
testThroughput/1/128/1         294 us          293 us            1 num_times_linearize_did_something=10 throughput=614.242M/s writes_per_iteration=11
testThroughput/0/4096/1        257 us          256 us            1 num_times_linearize_did_something=1 throughput=704.526M/s writes_per_iteration=12
testThroughput/1/4096/1        285 us          285 us            1 num_times_linearize_did_something=11 throughput=633.298M/s writes_per_iteration=11
testThroughput/0/1/2           283 us          283 us            1 num_times_linearize_did_something=11 throughput=637.848M/s writes_per_iteration=11
testThroughput/1/1/2           288 us          287 us            1 num_times_linearize_did_something=11 throughput=628.116M/s writes_per_iteration=11
testThroughput/0/128/2         316 us          316 us            1 num_times_linearize_did_something=7 throughput=571.135M/s writes_per_iteration=12
testThroughput/1/128/2         271 us          270 us            1 num_times_linearize_did_something=10 throughput=667.464M/s writes_per_iteration=11
testThroughput/0/4096/2        272 us          271 us            1 num_times_linearize_did_something=1 throughput=664.256M/s writes_per_iteration=13
testThroughput/1/4096/2        282 us          282 us            1 num_times_linearize_did_something=11 throughput=640.216M/s writes_per_iteration=11
testThroughput/0/1/3           283 us          282 us            1 num_times_linearize_did_something=11 throughput=639.23M/s writes_per_iteration=11
testThroughput/1/1/3           275 us          274 us            1 num_times_linearize_did_something=11 throughput=656.697M/s writes_per_iteration=11
testThroughput/0/128/3         281 us          280 us            1 num_times_linearize_did_something=5 throughput=642.652M/s writes_per_iteration=12
testThroughput/1/128/3         272 us          271 us            1 num_times_linearize_did_something=10 throughput=664.379M/s writes_per_iteration=11
testThroughput/0/4096/3        261 us          260 us            1 num_times_linearize_did_something=1 throughput=693.156M/s writes_per_iteration=14
testThroughput/1/4096/3        273 us          273 us            1 num_times_linearize_did_something=11 throughput=660.858M/s writes_per_iteration=11
```

Reply:
num_times_linearize_did_something looks high in the cases where maybeLinearize is used.  The number of copies done by both implementations is the same most of the time.

"if (slices_.size() >= 2 && slices_[1]->dataSize() >= max_size) {" should be using desired_min_size

Results after this change:
Benchmark                        Time             CPU   Iterations UserCounters...
testThroughput/0/0/0          60.5 us         60.5 us        11572 num_times_linearize_did_something=0 throughput=2.97744G/s writes_per_iteration=12
testThroughput/1/0/0          62.9 us         62.9 us        11116 num_times_linearize_did_something=10 throughput=2.86485G/s writes_per_iteration=11
testThroughput/0/1/1          62.1 us         62.1 us        11279 num_times_linearize_did_something=0 throughput=2.90119G/s writes_per_iteration=13
testThroughput/1/1/1          64.0 us         64.0 us        10919 num_times_linearize_did_something=11 throughput=2.81688G/s writes_per_iteration=11
testThroughput/0/128/1        64.4 us         64.4 us        10880 num_times_linearize_did_something=0 throughput=2.79695G/s writes_per_iteration=14
testThroughput/1/128/1        63.3 us         63.3 us        11091 num_times_linearize_did_something=10 throughput=2.84636G/s writes_per_iteration=11
testThroughput/0/4096/1       62.6 us         62.6 us        11181 num_times_linearize_did_something=0 throughput=2.87997G/s writes_per_iteration=13
testThroughput/1/4096/1       63.5 us         63.5 us        11035 num_times_linearize_did_something=11 throughput=2.83919G/s writes_per_iteration=11
testThroughput/0/1/2          60.9 us         60.9 us        11496 num_times_linearize_did_something=1 throughput=2.96085G/s writes_per_iteration=12
testThroughput/1/1/2          63.9 us         63.8 us        10956 num_times_linearize_did_something=11 throughput=2.82267G/s writes_per_iteration=11
testThroughput/0/128/2        63.3 us         63.2 us        11093 num_times_linearize_did_something=1 throughput=2.85011G/s writes_per_iteration=13
testThroughput/1/128/2        63.1 us         63.1 us        11066 num_times_linearize_did_something=10 throughput=2.85719G/s writes_per_iteration=11
testThroughput/0/4096/2       64.4 us         64.4 us        10890 num_times_linearize_did_something=0 throughput=2.80025G/s writes_per_iteration=14
testThroughput/1/4096/2       63.5 us         63.5 us        11029 num_times_linearize_did_something=11 throughput=2.83725G/s writes_per_iteration=11
testThroughput/0/1/3          61.0 us         60.9 us        11461 num_times_linearize_did_something=1 throughput=2.957G/s writes_per_iteration=12
testThroughput/1/1/3          64.0 us         64.0 us        10950 num_times_linearize_did_something=11 throughput=2.81811G/s writes_per_iteration=11
testThroughput/0/128/3        63.8 us         63.8 us        11085 num_times_linearize_did_something=1 throughput=2.82643G/s writes_per_iteration=13
testThroughput/1/128/3        63.1 us         63.0 us        11112 num_times_linearize_did_something=10 throughput=2.85861G/s writes_per_iteration=11
testThroughput/0/4096/3       64.7 us         64.7 us        10804 num_times_linearize_did_something=1 throughput=2.78513G/s writes_per_iteration=14
testThroughput/1/4096/3       63.3 us         63.3 us        11051 num_times_linearize_did_something=11 throughput=2.84711G/s writes_per_iteration=11